### PR TITLE
CoreUtilities.parseIndex added and applied to ListTag.get

### PR DESCRIPTION
**Pull request by TheHusseler**
Created a generic index parser that handles the following:
- Negative numbers subtract from the end of the list (i.e. ListTag.get[-1] will return the second to last item)
- get[first] and get[last] now supported, parsing to the first and last indexes of the list
- Numbers higher than the list size or lower than the negative list size will parse to the first or last index
- Indexes passed in will have -1 applied to them to match java zero-indexing

I've tested each edge case on my test server to ensure everything works.

### CODE REVIEW

File: src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java [hunk 1]
```diff
@@ -709,6 +709,28 @@ public static String getClosestOption(List<String> strs, String opt) {
...
+        } catch (Exception ex) {
+            return -1;
+        }
+        if(index == 0) {
+            return -1;
+        }
+        if (index < 0) {
+            return list.size() + Math.max(1 - list.size(), index) - 1;
+        }
+        return Math.min(list.size(), index) - 1;
```
Comment (acikek) [inline src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java:731]: Follow codebase formatting - space after `if`, if/else and try/catch blocks should have a newline separating them, e.g.
```java
if (cond) {

}
else {

}
```

Comment (TheHusseler) [reply src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java:731]: Ahh, I'll update those spaces, am I missing a spot where I didn't have the if/else separated by new line or is that just the catch you're referring to?

Comment (acikek) [reply src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java:731]: Just the catch, I provided if/else as it's the most common example.

File: src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java [hunk 2]
```diff
@@ -1420,22 +1423,20 @@ else if (input.size() > 1) {
                 if (indices.size() > 1) {
                     ListTag results = new ListTag();
                     for (String index : indices) {
-                        int ind = Integer.parseInt(index);
-                        if (ind > 0 && ind <= object.size()) {
-                            results.add(object.get(ind - 1));
+                        int ind = CoreUtilities.parseIndex(object.objectForms, index);
+                        if (ind >= 0) {
+                            results.add(object.get(ind));
+                        } else {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java:1429]: newline after `}`

File: src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java [hunk 3]
```diff
@@ -1420,22 +1423,20 @@ else if (input.size() > 1) {
...
-                    int index = Integer.parseInt(indices.get(0)) - 1;
-                    if (index >= object.size()) {
-                        attribute.echoError("Invalid list.get index '" + (index + 1) + "' ... list is only " + object.size() + " long.");
-                        return null;
-                    }
+                    int index = CoreUtilities.parseIndex(object.objectForms, indices.get(0));
                     if (index < 0) {
-                        attribute.echoError("Invalid list.get index '" + (index + 1) + "' ... must be at least 1.");
-                        index = 0;
+                        attribute.echoError("Invalid list.get index '" + (index + 1));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java:1438]: chopped  this error message weird

File: src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java [hunk 4]
```diff
@@ -1446,20 +1447,14 @@ else if (input.size() > 1) {
...
-                        else {
-                            index2 = attribute.getIntContext(2) - 1;
-                        }
-                        if (index2 >= object.size()) {
-                            index2 = object.size() - 1;
-                        }
-                        if (index2 < 0) {
-                            index2 = 0;
+                        int index2 = CoreUtilities.parseIndex(object.objectForms, attribute.getContext(2));
+                        if (index2 < 0 || index2 < index) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java:1455]: The `index2 < index` here is new and thus adds an error state that previously was undocumentedly valid. This should not be done unless there's a specific reason for it

File: src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java (hunk 3 above)
Comment (TheHusseler) [reply src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java:1438]: Chopped it since it can be negative now, really only handles the 0 case there. I'll update it to be specific to that

File: src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java [hunk 5]
```diff
@@ -709,6 +709,29 @@ public static String getClosestOption(List<String> strs, String opt) {
         return closest;
     }
 
+    public static int parseIndex(List list, String str) {
+        if (str.equalsIgnoreCase("last")) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java:713]: Many Java default methods like `equalsIgnoreCase` are avoided in Denizen due to performance issues, note the use of a `CoreUtilities` alternative in the original code you edited.
In this case as multiple checks are made, just do `str = CoreUtilities.toLowerCase(str);` followed by `.equals` calls

File: src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java (hunk 4 above)
Comment (TheHusseler) [reply src/main/java/com/denizenscript/denizencore/objects/core/ListTag.java:1455]: Fair point, I'd added it since it would always return an empty list but should have asked first. I'll remove it

File: src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java [hunk 6]
```diff
@@ -709,6 +709,29 @@ public static String getClosestOption(List<String> strs, String opt) {
...
+        }
+        if (str.equalsIgnoreCase("first")) {
+            return 0;
+        }
+        int index;
+        try {
+            index = Integer.parseInt(str);
+        }
+        catch (Exception ex) {
+            return -1;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java:724]: `get[0]` is not an error case, it should be equivalent to `get[1]`

File: src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java [hunk 7]
```diff
@@ -709,6 +709,29 @@ public static String getClosestOption(List<String> strs, String opt) {
...
+        }
+        if (index == 0) {
+            return -1;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/utilities/CoreUtilities.java:727]: `get[0]` is not an error case, it should be equivalent to `get[1]`
